src/diagnostics.rs

```diff
@@ -57,35 +57,77 @@ pub fn publish_diagnostics(params: Params, ctx: &mut Context) {
             )
         })
         .join(" ");
-    let mut lines_with_errors = HashSet::new();
-    let diagnostic_ranges = diagnostics
+
+    // Assemble a list of diagnostics by line number
+    let mut diagnostic_ranges = HashMap::new();
+    for x in diagnostics {
+        let line_number = x.range.end.line;
+
+        let face = match x.severity {
+            Some(DiagnosticSeverity::Error) => "InlayDiagnosticError",
+            _ => "InlayDiagnosticWarning",
+        };
+
+        let line_diagnostics = diagnostic_ranges
+            .entry(line_number)
+            .or_insert(LineDiagnostics {
+                range_end: x.range.end,
+                symbols: String::new(),
+                text: String::new(),
+                text_face: String::new(),
+                text_severity: None,
+            });
+
+        if let Some(severity) = x.severity {
```

Comment:
I am not happy with this code _at all_. This is what I would do in Lua, but it feels really weird in Rust. There has got to be a more idiomatic way to do these checks.

Reply:
Remembered `map_or`, which made this much cleaner. I knew there was a better solution somewhere.